In [5]:
import pandas as pd
from google.colab import files
up = files.upload()


Saving Day11.csv to Day11.csv


In [6]:
df = pd.read_csv("Day11.csv")
display(df.head())

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


In [8]:
print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())
print("\nData Types:")
print(df.dtypes)
print("\nMissing Values:")
print(df.isnull().sum())
print("\nDuplicate Records:", df.duplicated().sum())

Shape: (157, 12)

Columns:
['Employee_ID', 'Employee_Name', 'Department', 'Job_Title', 'Age', 'Gender', 'Annual_Salary', 'Experience_Years', 'Joining_Date', 'City', 'Performance_Score', 'Work_Mode']

Data Types:
Employee_ID           object
Employee_Name         object
Department            object
Job_Title             object
Age                  float64
Gender                object
Annual_Salary        float64
Experience_Years     float64
Joining_Date          object
City                  object
Performance_Score    float64
Work_Mode             object
dtype: object

Missing Values:
Employee_ID          0
Employee_Name        0
Department           5
Job_Title            0
Age                  4
Gender               5
Annual_Salary        5
Experience_Years     3
Joining_Date         0
City                 5
Performance_Score    3
Work_Mode            2
dtype: int64

Duplicate Records: 7


In [9]:
print("Rows containing at least one missing value:", df.isnull().any(axis=1).sum())

print("\nUnique Department values:")
print(df["Department"].value_counts(dropna=False))

print("\nUnique Gender values:")
print(df["Gender"].value_counts(dropna=False))

print("\nUnique City values:")
print(df["City"].value_counts(dropna=False))

print("\nUnique Work_Mode values:")
print(df["Work_Mode"].value_counts(dropna=False))

Rows containing at least one missing value: 32

Unique Department values:
Department
Marketing           23
Engineering         21
Customer Success    20
Finance             19
Sales               19
Operations          17
Human Resources     16
Data & Analytics    15
NaN                  5
ENGINEERING          1
engineering          1
Name: count, dtype: int64

Unique Gender values:
Gender
Male      75
Female    68
Other      7
NaN        5
female     1
MALE       1
Name: count, dtype: int64

Unique City values:
City
Chandigarh    24
Pune          23
Jaipur        18
Mumbai        15
Hyderabad     14
Srinagar      13
Bengaluru     12
Delhi         11
Chennai       10
Kolkata        9
NaN            5
delhi          1
DELHI          1
Delhi          1
Name: count, dtype: int64

Unique Work_Mode values:
Work_Mode
Remote    56
Hybrid    50
Office    47
NaN        2
REMOTE     1
remote     1
Name: count, dtype: int64


## 2. Cleaning Steps

- **drop_duplicates()** is utilized to eliminate exact duplicates in the rows.
- There is standardization of Department, Gender, City, and Work_Mode formats.
- **Joining_Date** is converted from an object/string to Pandas datetime format.
- For numeric variables, missing values are filled with the median since median value is less affected by phenomena like high salaries, ages, experience, or scores.
- For categorical variables, missing values are filled with a mode as they are nominal in nature.
- **dropna()** method is not applied for the main cleaning purpose as deletion of rows means loss of useful employee records.
- Forward Filling is not applied since all employee information is independent records; thus carrying the value of one employee over another employee is not acceptable.

In [10]:
cleaned = df.copy()
cleaned = cleaned.drop_duplicates()
cleaned["Department"] = cleaned["Department"].replace({
    "ENGINEERING": "Engineering",
    "engineering": "Engineering"
})
cleaned["Gender"] = cleaned["Gender"].replace({
    "MALE": "Male",
    "female": "Female"
})
cleaned["City"] = cleaned["City"].astype("string").str.strip()
cleaned["City"] = cleaned["City"].replace({
    "DELHI": "Delhi",
    "delhi": "Delhi"
})
cleaned["Work_Mode"] = cleaned["Work_Mode"].replace({
    "REMOTE": "Remote",
    "remote": "Remote"
})
cleaned["Joining_Date"] = pd.to_datetime(
    cleaned["Joining_Date"],
    errors="coerce"
)
numeric_cols = [
    "Age",
    "Annual_Salary",
    "Experience_Years",
    "Performance_Score"
]
for col in numeric_cols:
    cleaned[col] = cleaned[col].fillna(cleaned[col].median())
categorical_cols = [
    "Department",
    "Gender",
    "City",
    "Work_Mode"
]
for col in categorical_cols:
    cleaned[col] = cleaned[col].fillna(cleaned[col].mode()[0])
cleaned["Age"] = cleaned["Age"].astype(int)
cleaned["Performance_Score"] = cleaned["Performance_Score"].astype(int)
display(cleaned.head())

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Engineering,Data Scientist,48,Other,108371.0,13.2,2017-06-13,Pune,5,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31,Female,108824.0,12.1,2021-09-15,Mumbai,2,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28,Female,119400.0,13.2,2021-02-05,Bengaluru,3,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30,Female,111847.0,9.2,2018-08-05,Jaipur,4,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25,Female,78677.0,10.9,2018-07-24,Hyderabad,4,Hybrid


In [11]:
print("Cleaned Shape:", cleaned.shape)

print("\nMissing Values After Cleaning:")
print(cleaned.isnull().sum())

print("\nDuplicate Records After Cleaning:", cleaned.duplicated().sum())

print("\nData Types After Cleaning:")
print(cleaned.dtypes)

Cleaned Shape: (150, 12)

Missing Values After Cleaning:
Employee_ID          0
Employee_Name        0
Department           0
Job_Title            0
Age                  0
Gender               0
Annual_Salary        0
Experience_Years     0
Joining_Date         0
City                 0
Performance_Score    0
Work_Mode            0
dtype: int64

Duplicate Records After Cleaning: 0

Data Types After Cleaning:
Employee_ID                  object
Employee_Name                object
Department                   object
Job_Title                    object
Age                           int64
Gender                       object
Annual_Salary               float64
Experience_Years            float64
Joining_Date         datetime64[ns]
City                 string[python]
Performance_Score             int64
Work_Mode                    object
dtype: object


In [12]:
print("Department values after cleaning:")
print(sorted(cleaned["Department"].unique()))

print("\nGender values after cleaning:")
print(sorted(cleaned["Gender"].unique()))

print("\nCity values after cleaning:")
print(sorted(cleaned["City"].unique()))

print("\nWork_Mode values after cleaning:")
print(sorted(cleaned["Work_Mode"].unique()))

Department values after cleaning:
['Customer Success', 'Data & Analytics', 'Engineering', 'Finance', 'Human Resources', 'Marketing', 'Operations', 'Sales']

Gender values after cleaning:
['Female', 'Male', 'Other']

City values after cleaning:
['Bengaluru', 'Chandigarh', 'Chennai', 'Delhi', 'Hyderabad', 'Jaipur', 'Kolkata', 'Mumbai', 'Pune', 'Srinagar']

Work_Mode values after cleaning:
['Hybrid', 'Office', 'Remote']


In [13]:
comparison = pd.DataFrame({
    "Metric": [
        "Rows",
        "Columns",
        "Missing Cells",
        "Duplicate Rows"
    ],
    "Before Cleaning": [
        df.shape[0],
        df.shape[1],
        int(df.isnull().sum().sum()),
        int(df.duplicated().sum())
    ],
    "After Cleaning": [
        cleaned.shape[0],
        cleaned.shape[1],
        int(cleaned.isnull().sum().sum()),
        int(cleaned.duplicated().sum())
    ]
})

display(comparison)

,Metric,Before Cleaning,After Cleaning
0,Rows,157,150
1,Columns,12,12
2,Missing Cells,32,0
3,Duplicate Rows,7,0


In [14]:
cleaned.to_csv(
    "Cleaned_Company_Employee_Dataset.csv",
    index=False
)

print("Cleaned dataset exported successfully.")

Cleaned dataset exported successfully.


## Cleaning Summary

The original data comprised of **157 records with 12 columns**. Initially, it had a total of **32 missing cells** and **7 duplicate records**. The columns had **inconsistent capitalizations** in certain categorical c**olumns, which includes Department, - Gender, City, and Work_Mode**. The **missing numeric values were replaced using median imputation and missing categorical values by mode imputation**. The **Joining_Date column was converted from object/string column type to datetime**. After the cleaning process, the **cleaned data consists of 150 unique employee records** with no missing values or duplicate records. The cleaned dataset is saved in a file named Cleaned_Company_Employee_Dataset.csv.